### PR TITLE
feat: ignore IIFE's in the `no-loop-func` rule

### PR DESCRIPTION
**Pull request by snitin315**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[x] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)
Fixes https://github.com/eslint/eslint/issues/16902

```js
var arr = [];

for (var i = 0; i < 5; i++) {
    arr.push((f => f)((() => i)())); // OK
}

for (var i = 0; i < 5; i++) {
    arr.push((() => {
        return (() => i)(); // OK
    })());
}

for (var i = 0; i < 5; i++) {
    arr.push((f => f)(() => i)); // Error
}

for (var i = 0; i < 5; i++) {
    arr.push((() => {
        return () => i; // Error
    })());
}

for (var i = 0; i < 5; i++) {
    (function fun () {
        if (arr.includes(fun)) return i; // Error
        else arr.push(fun);
    })();
}
```

### CODE REVIEW

File: docs/src/rules/no-loop-func.md
Comment (nzakas) [inline docs/src/rules/no-loop-func.md:91]: Is there a reason you used `var` in these examples instead of `let` as in previous ones?

**Review by nzakas — CHANGES_REQUESTED** [review]
Code LGTM. Can you also add a note in the rule details that explains that IIFEs are ignored while generators/async are not?

File: lib/rules/no-loop-func.js
Comment (mdjermanovic) [inline lib/rules/no-loop-func.js:209]: This checks some very specific cases only. For example, this would be a false negative:

```js
/* eslint no-loop-func: "error" */

var arr = [];

for (var i = 0; i < 5; i++) {
    arr.push((() => {
        return () => { return i }; // false negative
    })());
}

console.log(arr.map(f => f())); // [ 5, 5, 5, 5, 5 ]

```

If we want to check if the reference is not in a non-IIFE inner function, we could use `r.from.variableScope.block` to get the function node where the reference is, and then check if it's an IIFE. We should also check its parent functions up to the current one (`node`).

File: docs/src/rules/no-loop-func.md
Comment (snitin315) [reply docs/src/rules/no-loop-func.md:91]: Yeah, it works as expected with `let`.

with `var` it would return [5, 5, 5, 5, 5] with let: `[0, 1, 2, 3, 4]`

File: lib/rules/no-loop-func.js
Comment (snitin315) [reply lib/rules/no-loop-func.js]: This is not required, I had added it initially while testing something. Can be removed safely.

File: lib/rules/no-loop-func.js
Comment (snitin315) [reply lib/rules/no-loop-func.js]: I've updated the code, maybe we don't need to check parent functions upto current `node`, [this one seems false positive](https://github.com/eslint/eslint/actions/runs/6130619171/job/16639879414?pr=17528#step:5:56):
```js
/* eslint no-loop-func: "error" */

 var arr = [];

for (var i = 0; i < 5; i++) {
    arr.push((() => {
        return (() => i)(); // false posiitve
    })());
}

console.log(arr.map(f => f))  // [ 0, 1, 2, 3, 4]

```

File: lib/rules/no-loop-func.js
Comment (mdjermanovic) [inline lib/rules/no-loop-func.js:227]: We don't need to check if every ancestor node is an IIFE, just if every separate execution context (e.g., an inner function) is immediately executed. If the innermost is `r.from.variableScope`, the next upper would be `r.from.variableScope.upper.variableScope` and so on.

File: lib/rules/no-loop-func.js
Comment (mdjermanovic) [reply lib/rules/no-loop-func.js]: > I've updated the code, maybe we don't need to check parent functions upto current `node`, [this one seems false positive](https://github.com/eslint/eslint/actions/runs/6130619171/job/16639879414?pr=17528#step:5:56):

Left a comment on the relevant line: https://github.com/eslint/eslint/pull/17528#discussion_r1320621371

File: lib/rules/no-loop-func.js
Comment (snitin315) [reply lib/rules/no-loop-func.js]: Updated, thanks!

File: docs/src/rules/no-loop-func.md
Comment (nzakas) [inline docs/src/rules/no-loop-func.md:37]: We need to spell out what IIFE means because it wasn't mentioned earlier. Also cleaning up the language a bit.

```suggestion
This rule disallows any function within a loop, including async functions and generators, that contains unsafe references (e.g. to modified variables from the outer scope). This rule ignores immediately-invoked function expressions (IIFEs) because they do not persist outside of the loop and are therefore always safe.
```

**Review by nzakas — APPROVED** [review]
LGTM. Just a suggestion clean up the docs.

File: lib/rules/no-loop-func.js
Comment (mdjermanovic) [inline lib/rules/no-loop-func.js:205]: We should also check if inner functions are not async/generators.

```js
/* eslint no-loop-func: "error" */

var arr = [];

for (var i = 0; i < 5; i ++) {
    // false negative, reference to `i` is unsafe
    (() => {
        arr.push((async () => {
            await 1;
            return i;
        })());
    })();
}

(async () => {
    console.log(await Promise.all(arr)); // [ 5, 5, 5, 5, 5 ]
})();

```

File: lib/rules/no-loop-func.js
Comment (mdjermanovic) [inline lib/rules/no-loop-func.js:213]: We should check this for all inner functions.

```js
/* eslint no-loop-func: "error" */

var arr = [];

for (var i = 0; i < 5; i ++) {
    // false negative, reference to `i` is unsafe
    (() => {
        (function f() {
            if (!arr.includes(f)) {
                arr.push(f);
            }
            return i;
        })();
    })();

}

console.log(arr.map(f => f())); // [ 5, 5, 5, 5, 5 ]

```

File: lib/rules/no-loop-func.js
Comment (mdjermanovic) [inline lib/rules/no-loop-func.js:228]: This checks if the `node` itself is an IIFE only if the reference is directly in it. Since the `if (parentNode.type === "CallExpression" && (isFunctionExpression || ArrowFunctionExpression))` check is incomplete, there will be false negatives for references in inner functions when `node` is not IIFE but an argument of another IIFE:

```js
/* eslint no-loop-func: "error" */

var arr = [];

for (var i = 0; i < 5; i ++) {
    // false negative, reference to `i` is unsafe
    ((f) => {
        arr.push(f);
    })(() => {
        return (() => i)();
    });

}

console.log(arr.map(f => f())); // [ 5, 5, 5, 5, 5 ]

```

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Comment by mdjermanovic** [comment]
I think we could simplify this a lot by taking a different approach, that is to skip IIFEs (that are not async, generators, or self-referencing) here and thus let the rule check and report their inner functions directly.

https://github.com/eslint/eslint/blob/bd7a71fd6b7efb0445393304e2d48c5c06d46a45/lib/rules/no-loop-func.js#L47-L52

For example, the current version of this rule reports error for unsafe reference to `i` on the IIFE:

```js
/* eslint no-loop-func: "error" */

var arr = [];

for (var i = 0; i < 5; i++) {
    arr.push((
        () => { // <-- error here
            return () => i;
        }
    )());
}

```

After this change, it would report the same error directly on the inner function:

```js
/* eslint no-loop-func: "error" */

var arr = [];

for (var i = 0; i < 5; i++) {
    arr.push((
        () => {
            return () => i; // <-- error here
        }
    )());
}

```

`function checkForLoops` could just skip IIFEs without checking anything about their inner functions since they'll be checked directly later in the traversal.

The only downside is that this can change error locations (and thus existing `eslint-disable-next?-line` comments would have to be reallocated) and the total number of errors if there are multiple inner functions, but I think it's okay for a semver-minor release.

**Comment by Rec0iL99** [comment]
Not Stale. @snitin315 is less active at the moment. Will be back soon.

**Comment by snitin315** [comment]
>The only downside is that this can change error locations (and thus existing `eslint-disable-next?-line` comments would have to be reallocated) and the total number of errors if there are multiple inner functions, but I think it's okay for a semver-minor release.

@mdjermanovic Once we changed error locations for `max-lines-per-function`, we had to revert it later in https://github.com/eslint/eslint/pull/15397. So, I'm not sure if we should do it in a minor release here.

**Comment by mdjermanovic** [comment]
> > The only downside is that this can change error locations (and thus existing `eslint-disable-next?-line` comments would have to be reallocated) and the total number of errors if there are multiple inner functions, but I think it's okay for a semver-minor release.
> 
> @mdjermanovic Once we changed error locations for `max-lines-per-function`, we had to revert it later in #15397. So, I'm not sure if we should do it in a minor release here.

I think it's okay in this case. For `max-lines-per-function`, the change wasn't a bug fix and it was affecting every lint problem reported by this rule, while this change is considered a bug fix and affects only problems on IIFEs, and most eslint-disable comments on IIFEs should be just removed so I wouldn't expect many cases where eslint-disable comments need to be reallocated (Loop > IIFE with inner functions seems like a rare case).

**Comment by snitin315** [comment]
Makes sense, I'll update the PR 👍🏻

**Comment by Rec0iL99** [comment]
Ping @snitin315

File: lib/rules/no-loop-func.js
Comment (snitin315) [reply lib/rules/no-loop-func.js:227]: Good catch, fixed!

File: lib/rules/no-loop-func.js
Comment (snitin315) [reply lib/rules/no-loop-func.js:203]: Thanks for the catch, fixed.

**Comment by snitin315** [comment]
@mdjermanovic Thanks for the detailed review, I've resolved and added corresponding test cases to your comments.

File: lib/rules/no-loop-func.js
Comment (snitin315) [reply lib/rules/no-loop-func.js:247]: Apologies for the delayed respose. Do you mean we can remove this whole logic?

Comment (mdjermanovic) [reply lib/rules/no-loop-func.js:247]: Yes, I think we can remove all code that handles inner functions of skipped ones as a special case. But we should also update the code that checks if a function is already in a function that is in a loop (to avoid duplicate reports) to also skip the same functions.

**Comment by mdjermanovic** [comment]
This example should now be moved into the examples of correct code:

https://github.com/eslint/eslint/blob/389744be255717c507fafc158746e579ac08d77e/docs/src/rules/no-loop-func.md?plain=1#L44-L46

File: lib/rules/no-loop-func.js
Comment (mdjermanovic) [inline lib/rules/no-loop-func.js:202]: We can move this after `if (!(node.async || node.generator)) {}` as there's no need to find unsafe references if the function is going to be skipped.

File: tests/lib/rules/no-loop-func.js
Comment (mdjermanovic) [inline tests/lib/rules/no-loop-func.js:210]:
```suggestion
            languageOptions: { ecmaVersion: 6 }
```

This is in the `valid` array.

File: lib/rules/no-loop-func.js
Comment (mdjermanovic) [inline lib/rules/no-loop-func.js:63]: Since not all IIFEs are skipped, this could produce duplicate errors:

```js
/* eslint no-loop-func: 2 */

for (var i = 0; i < 5; i++) {
    (async () => {
        () => i;
    })();
}
```

```
  4:6  error  Function declared in a loop contains unsafe references to variable(s) 'i'  no-loop-func
  5:9  error  Function declared in a loop contains unsafe references to variable(s) 'i'  no-loop-func
```

In this case, only the error on the upper function (line 4) should be reported, as before this change ([Playground demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLWxvb3AtZnVuYzogMiAqL1xuXG5mb3IgKHZhciBpID0gMDsgaSA8IDU7IGkrKykge1xuICAgIChhc3luYyAoKSA9PiB7XG4gICAgICAgICgpID0+IGk7XG4gICAgfSkoKTtcbn0iLCJvcHRpb25zIjp7InJ1bGVzIjp7fSwibGFuZ3VhZ2VPcHRpb25zIjp7InNvdXJjZVR5cGUiOiJtb2R1bGUiLCJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19)).

Perhaps we could maintain the list of skipped functions in a set, and check here if the functions is in that set.

Comment (snitin315) [reply lib/rules/no-loop-func.js]: Good catch! Fixed.

File: lib/rules/no-loop-func.js
Comment (mdjermanovic) [inline lib/rules/no-loop-func.js:63]:
```suggestion
                if (SKIPPED_IIFE_NODES.has(parent)) {
```

There's no need to check again if it's an IIFE when it's skipped as an IIFE.

File: lib/rules/no-loop-func.js
Comment (mdjermanovic) [inline lib/rules/no-loop-func.js:12]: Since this Set is at the module level, it will hold references to the nodes, and indirectly to the entire AST which therefore cannot be removed from memory even after the file's linting is complete and its AST is no longer needed.

To avoid memory leaks, I think it's fine to move this Set and helper functions that are using it into `create(context) {}`.

Comment (snitin315) [reply lib/rules/no-loop-func.js]: Updated 👍🏻

File: tests/lib/rules/no-loop-func.js
Comment (mdjermanovic) [inline tests/lib/rules/no-loop-func.js:455]:
```suggestion
            errors: [{ messageId: "unsafeRefs", data: { varNames: "'i'" }, type: "ArrowFunctionExpression", line: 6 }]
```

File: tests/lib/rules/no-loop-func.js
Comment (mdjermanovic) [inline tests/lib/rules/no-loop-func.js:488]:
```suggestion
           errors: [{ messageId: "unsafeRefs", data: { varNames: "'j'" }, type: "ArrowFunctionExpression", line: 7 }]
```

File: tests/lib/rules/no-loop-func.js
Comment (mdjermanovic) [inline tests/lib/rules/no-loop-func.js:504]:
```suggestion
            errors: [{ messageId: "unsafeRefs", data: { varNames: "'i'" }, type: "ArrowFunctionExpression", line: 7 }]
```

File: tests/lib/rules/no-loop-func.js
Comment (mdjermanovic) [inline tests/lib/rules/no-loop-func.js:515]:
```suggestion
            errors: [{ messageId: "unsafeRefs", data: { varNames: "'i'" }, type: "ArrowFunctionExpression", line: 3 }]
```

**Review by mdjermanovic — COMMENTED** [review]
Just a few suggestions about the tests, then LGTM.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

Since this is marked as a feature, it needs another approval from a team member before merging.

**Pull request merged**
